### ISSUE TITLE
The tree adapter is not used for updating the end location of an element

### ISSUE THREAD

**Issue by DMartens** [inikulin/parse5#314]
Hi,
I am currently writing a custom tree adapter restructuring the source code location:
```javascript
{
	start: { offset: 0, line: 1, column: 1 },
	end:   { offset: 0, line: 1, column: 1 },
}
```
The problem is that elements closing tag are not correctly handled in the [parser mixin](https://github.com/inikulin/parse5/blob/252819607421a5741cf745bb60c404f023531b0d/packages/parse5/lib/extensions/location-info/parser-mixin.js#L35).
The location is retrieved using the tree adapter but the end location is set directly on the retrieved location, not using the tree adapter.
Therefore the result after parsing is:
```javascript
{
	start: { offset: 0, line: 1, column: 1 },
	end:   { offset: 0, line: 1, column: 1 },
	endTag: ...,
    endLine: 1,
	endCol: 10,
	endOffset: 9,
}
```

I think there are three possible solutions:
1. Call the `setSourceCodeLocation` only after the closing tag is parsed
2. Call the `setSourceCodeLocation` twice (once after the opening tag and once for the closing tag)
3. Introduce a new method `updateSourceCodeLocation` for this purpose

Personally I would prefer the last option.
If this change is approved, I gladly submit a PR for the proposed solution.

**Comment by inikulin** [comment]
Hi,

The original idea was that user don't have control over the source location structure. Therefore, `set/getSourceCodeLocation` methods should be used solely for data storage/retrievement and keep location data itself intact. 

This design was introduced because basically I didn't expect that someone would need custom location data representation. However, if it's a case for you, I'm not oppose to introduction of `updateSourceCodeLocation` and would welcome a PR for that. Though, not that would be a breaking change and will require a major version bump and proper documentation.

**Comment by inikulin** [comment]
Closed in #315
